File: C++/BFS/BFS-Maze.cpp
#include<iostream>
#include<queue>
using namespace std;

typedef pair<int,int> pii;
queue<pii> q;

// s = start, e = end
// 0 = path, 1 = wall
char maze[7][8]= {
{'s','0','1','1','1','1','1','1'},
{'1','0','0','0','1','0','0','1'},
{'1','1','0','1','0','0','1','1'},
{'1','1','0','1','0','0','0','1'},
{'1','0','0','0','0','1','0','e' },
{'1','1','0','1','1','1','1','1'},
};

// delta y, x
int dy[4] = {0,0,-1,1};
int dx[4] = {-1,1,0,0};
bool visited[7][8];

int cnt;
int flag;

int OOB(int y, int x);

void maze_print()
{
    cout <<"=step"<<cnt<<"=\n";
    for(int i=0; i<7; i++)
    {
        for(int j=0; j<8; j++)
            cout << maze[i][j];
        cout <<"\n";
    }
}

void bfs(int row, int col)
{
    
    maze[row][col] = 'x';
    maze_print();
    
    visited[row][col] = true;
    q.push({row, col});
    
    while(!q.empty())
    {
        pii cur = q.front(); q.pop();
        
        for(int dir=0; dir<4;dir++)
        {
            int new_row = cur.first + dy[dir];
            int new_col = cur.second + dx[dir];
            
            if(OOB(new_row, new_col))
                continue;
            
            if (maze[new_row][new_col] == 'e')
            {
                cout << "find the exit!!\n";
                cout << "Search: "<< cnt<<" times\n";
                exit(0);
            }
            
            if(!visited[new_row][new_col] && maze[new_row][new_col] == '0')
            {
                cnt++;
                maze[new_row][new_col] = 'x';
                maze_print();
                
                q.push({new_row, new_col});
            }
        }
    }
}
    
    
int main()
{
    bfs(0,0);
    
    if(!flag)
    {
        cout << "Couldn't find the exit!!\n";
        cout << "Search: "<< cnt<<" times\n";
    }
}

// Out Of Boundary
int OOB(int y, int x)
{
    return y<0 || y>=7 || x<0 || x>=8;
}

File: Python/BFS/BFS-Maze.py
from collections import deque


def out_of_boundary(row, col):
    if row < 0 or row >= 6 or col < 0 or col >= 8:
        return True
    return False


def maze_print():
    print("===step{}====".format(cnt))
    for i in range(6):
        for j in range(8):
            print(maze[i][j], end=" ")
        print("")


def bfs(row, col):
    global maze, visited, cnt

    q = deque()

    maze[row][col] = 'x'
    visited[row][col] = True
    q.append((row, col))

    maze_print()

    while len(q) > 0:
        cur_row, cur_col = q.popleft()

        for dir in range(4):
            new_row = cur_row + dy[dir]
            new_col = cur_col + dx[dir]

            if out_of_boundary(new_row, new_col):
                continue

            if maze[new_row][new_col] == 'e':
                print("Find the exit!")
                print("Search {} times".format(cnt))
                exit(0)

            if not visited[new_row][new_col] and maze[new_row][new_col] == '0':
                cnt += 1
                maze[new_row][new_col] = 'x'
                maze_print()

                q.append((new_row, new_col))


if __name__ == '__main__':
    maze = [['s', '0', '1', '1', '1', '1', '1', '1'],
            ['1', '0', '0', '0', '1', '0', '0', '1'],
            ['1', '1', '0', '1', '0', '0', '1', '1'],
            ['1', '1', '0', '1', '0', '0', '0', '1'],
            ['1', '0', '0', '0', '0', '1', '0', 'e'],
            ['1', '1', '0', '1', '1', '1', '1', '1']]

    visited = [[False for _ in range(8)] for _ in range(7)]

    # delta y, x
    dy = [0, 0, -1, 1]
    dx = [-1, 1, 0, 0]

    cnt = 1

    bfs(0, 0)